sklearn_numba_dpex/kmeans/engine.py

```diff
@@ -156,155 +205,170 @@ def kmeans_single(self, X, sample_weight, centers_init):
+            device = dpctl.SyclDevice(self.device)
+        elif isinstance(X, dpt.usm_ndarray):
+            device = X.device.sycl_device
+        else:
+            device = dpctl.SyclDevice()
 
+        accepted_dtypes = [np.float32]
+        # NB: one could argue that `float32` is a better default, but sklearn defaults
+        # to `np.float64` and we apply the same for consistency.
+        if device.has_aspect_fp64:
+            accepted_dtypes = [np.float64, np.float32]
+        else:
+            accepted_dtypes = [np.float32]
+
+        with _validate_with_array_api(device):
+            try:
+                X = self.estimator._validate_data(
+                    X,
+                    accept_sparse=False,
+                    dtype=accepted_dtypes,
+                    order=self.order,
+                    copy=False,
+                    reset=reset,
+                    force_all_finite=True,
+                    estimator=self.estimator,
+                )
+                return X
+            except TypeError as type_error:
+                if "A sparse matrix was passed, but dense data is required" in str(
+                    type_error
+                ):
+                    raise NotSupportedByEngineError from type_error
+
+    def _check_sample_weight(self, sample_weight, X):
+        """Adapted from sklearn.utils.validation._check_sample_weight to be compatible
+        with Array API dispatch"""
+        n_samples = X.shape[0]
+        dtype = X.dtype
+        device = X.device.sycl_device
         if sample_weight is None:
-            sample_weight = np.ones(len(X), dtype=X.dtype)
-
-        X, sample_weight, cluster_centers, output_dtype = self._set_dtype(
-            X, sample_weight, cluster_centers
-        )
-
-        return X, sample_weight, cluster_centers, output_dtype
-
-    def _set_dtype(self, X, sample_weight, cluster_centers):
-        output_dtype = compute_dtype = np.dtype(X.dtype).type
-        copy = True
-        if (compute_dtype != np.float32) and (compute_dtype != np.float64):
-            text = (
-                f"KMeans has been set to compute with type {compute_dtype} but only "
-                f"the types float32 and float64 are supported. The computations and "
-                f"outputs will default back to float32 type."
-            )
-            output_dtype = compute_dtype = np.float32
-        elif (compute_dtype == np.float64) and not self.device.has_aspect_fp64:
-            text = (
-                f"KMeans is set to compute with type {compute_dtype} but this type is "
-                f"not supported by the device {self.device.name}. The computations "
-                f"will default back to float32 type."
-            )
-            compute_dtype = np.float32
-
+            sample_weight = dpt.ones(n_samples, dtype=dtype, device=device)
+        elif isinstance(sample_weight, numbers.Number):
+            sample_weight = dpt.full(n_samples, 1, dtype=dtype, device=device)
```

Comment:
What are the reasons not to have only one branch here?

Reply:
A mistake! the latter line should read:

```suggestion
            sample_weight = dpt.ones(n_samples, dtype=dtype, device=device)
        elif isinstance(sample_weight, numbers.Number):
            sample_weight = dpt.full(n_samples, sample_weight, dtype=dtype, device=device)
```